File: java/src/main/java/io/cucumber/prettyformatter/ExceptionFormatter.java
package io.cucumber.prettyformatter;

import io.cucumber.messages.types.Exception;
import io.cucumber.messages.types.TestStepResultStatus;
import org.jspecify.annotations.Nullable;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.StringReader;
import java.util.Optional;

import static io.cucumber.prettyformatter.Theme.Element.STEP;

final class ExceptionFormatter {

    private final int indent;
    private final Theme theme;
    private final TestStepResultStatus status;

    ExceptionFormatter(int indent, Theme theme, TestStepResultStatus status) {
        this.indent = indent;
        this.theme = theme;
        this.status = status;
    }

    Optional<String> format(Exception exception, @Nullable String standaloneMessage) {
        // For FAILED, prefer stack trace, fall back to message
        if (status == TestStepResultStatus.FAILED) {
            if (exception.getStackTrace().isPresent()) {
                return Optional.of(format(exception.getStackTrace().get()));
            }
            if (exception.getMessage().isPresent()) {
                return Optional.of(format(exception.getMessage().get()));
            }
            return Optional.ofNullable(standaloneMessage).map(this::format);
        }

        // For PENDING/SKIPPED, only show the message (not stack trace)
        if (status == TestStepResultStatus.PENDING || status == TestStepResultStatus.SKIPPED) {
            if (exception.getMessage().isPresent()) {
                return Optional.of(format(exception.getMessage().get()));
            }
            return Optional.ofNullable(standaloneMessage).map(this::format);
        }

        // For all other statuses, return nothing
        return Optional.empty();
    }

    Optional<String> format(Exception exception) {
        return format(exception, null);
    }

    String format(String message) {
        LineBuilder builder = new LineBuilder(theme);
        // Read the lines in the message and add extra indentation
        try (BufferedReader lines = new BufferedReader(new StringReader(message))) {
            String line;
            while ((line = lines.readLine()) != null) {
                builder.indent(indent)
                        .append(STEP, status, line)
                        .newLine();
            }
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
        return builder
                .build();
    }
}

File: cpp/src/cucumber/pretty-formatter/ExceptionFormatter.cpp
#include "cucumber/pretty-formatter/ExceptionFormatter.hpp"
#include "cucumber/messages/Exception.hpp"
#include "cucumber/messages/TestStepResultStatus.hpp"
#include "cucumber/pretty-formatter/LineBuilder.hpp"
#include "cucumber/pretty-formatter/Theme.hpp"
#include <cstddef>
#include <memory>
#include <optional>
#include <sstream>
#include <string>
#include <utility>

namespace cucumber::pretty_formatter
{
    ExceptionFormatter::ExceptionFormatter(std::size_t indent, std::shared_ptr<Theme> theme, messages::TestStepResultStatus status)
        : indent{ indent }
        , theme{ std::move(theme) }
        , status{ status }
    {}

    std::optional<std::string> ExceptionFormatter::Format(const std::shared_ptr<const messages::Exception>& exception,
        const std::optional<std::string>& message)
    {
        if (status == messages::TestStepResultStatus::FAILED && exception->stackTrace.has_value())
        {
            return Format(exception->stackTrace.value());
        }

        if (status == messages::TestStepResultStatus::FAILED || status == messages::TestStepResultStatus::PENDING ||
            status == messages::TestStepResultStatus::SKIPPED)
        {
            if (exception->message.has_value())
            {
                return Format(exception->message.value());
            }
            if (message.has_value())
            {
                return Format(message.value());
            }
            return std::nullopt;
        }

        return std::nullopt;
    }

    std::optional<std::string> ExceptionFormatter::Format(const std::shared_ptr<const messages::Exception>& exception)
    {
        return Format(exception, std::nullopt);
    }

    std::string ExceptionFormatter::Format(const std::string& message)
    {
        LineBuilder builder{ theme };

        std::istringstream stream{ message };
        for (std::string line; std::getline(stream, line);)
        {
            builder.Indent(indent).Append(Theme::Element::step, status, line).NewLine();
        }

        return builder.Build();
    }
}
